include/boost/geometry/algorithms/simplify.hpp

```diff
@@ -116,34 +117,174 @@ struct simplify_range
+        std::size_t i = 0;
+        for (typename boost::range_iterator<Ring const>::type
+                it = boost::begin(ring); it != boost::end(ring); ++it, ++i)
+        {
+            // This actually is point-segment distance but will result
+            // in point-point distance
+            typename Strategy::distance_type dist = distance_strategy.apply(*it, point, point);
+            if (dist > max_distance)
+            {
+                max_distance = dist;
+                index = i;
+            }
+        }
+        return index;
+    }
+
+public :
+    template <typename Ring, typename Strategy, typename Distance>
+    static inline void apply(Ring const& ring, Ring& out,
+                    Distance const& max_distance, Strategy const& strategy)
+    {
+        std::size_t const size = boost::size(ring);
+        if (size == 0)
+        {
+            return;
+        }
+
+        int const input_sign = area_sign(geometry::area(ring));
+
+        std::set<std::size_t> visited_indexes;
+
+        // Rotate it into a copied vector
+        // (vector, because source type might not support rotation)
+        // (duplicate end point will be simplified away)
+        typedef typename geometry::point_type<Ring>::type point_type;
+
+        std::vector<point_type> rotated(size);
+
+        // Closing point (but it will not start here)
+        std::size_t index = 0;
+
+        // Iterate (usually one iteration is enough)
+        for (std::size_t iteration = 0; iteration < 4u; iteration++)
+        {
+            // Always take the opposite. Opposite guarantees that no point
+            // "halfway" is chosen, creating an artefact (very narrow triangle)
+            // Iteration 0: opposite to closing point (1/2, = on convex hull)
+            //              (this will start simplification with that point
+            //               and its opposite ~0)
+            // Iteration 1: move a quarter on that ring, then opposite to 1/4
+            //              (with its opposite 3/4)
+            // Iteration 2: move an eight on that ring, then opposite (1/8)
+            // Iteration 3: again move a quarter, then opposite (7/8)
+            // So finally 8 "sides" of the ring have been examined (if it were
+            // a semi-circle). Most probably, there are only 0 or 1 iterations.
+            switch (iteration)
+            {
+                case 1 : index = (index + size / 4) % size; break;
+                case 2 : index = (index + size / 8) % size; break;
+                case 3 : index = (index + size / 4) % size; break;
+            }
+            index = get_opposite<Strategy>(index, ring);
+
+            if (visited_indexes.count(index) > 0)
+            {
+                // Avoid trying the same starting point more than once
+                continue;
+            }
+
+            std::rotate_copy(boost::begin(ring), boost::begin(ring) + index,
+                             boost::end(ring), rotated.begin());
```

Comment:
`boost::begin(ring) + index` could be replaced with `range::pos(ring, index)`. My guess is that because index is of unsigned type and the random access iterator is advanced with difference_type which is signed there can be a conversion warning generated here.